File: app/src/main/java/com/tohsoft/app/ui/history/HistoryActivity.java
package com.tohsoft.app.ui.history;

import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.tohsoft.app.R;
import com.tohsoft.app.databinding.ActivityHistoryBinding;
import com.tohsoft.base.mvp.ui.BaseActivity;
import com.tohsoft.base.mvp.ui.BasePresenter;

import java.util.ArrayList;
import java.util.List;


/**
 * Created by PhongNX on 2/10/2020.
 * <p>
 * Template cho việc hiển thị list data (hiển thị emptyAd...)
 * <p>
 * Note: Nếu không dùng Activity này thì xóa đi
 */
public class HistoryActivity extends BaseActivity<HistoryMvpPresenter> implements HistoryMvpView, View.OnClickListener {
    private ActivityHistoryBinding mBinding;
    private HistoryItemAdapter mAdapter;

    @Override
    protected BasePresenter onRegisterPresenter() {
        return new HistoryPresenter(getContext());
    }

    @Override
    protected ViewGroup getBottomAdsContainer() {
        return mBinding.frBottomBanner;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = ActivityHistoryBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());

        init();

        mPresenter.getData();
    }

    private void init() {
        mAdapter = new HistoryItemAdapter(mContext, new ArrayList<>());
        mBinding.rvData.setLayoutManager(new LinearLayoutManager(mContext));
        mBinding.rvData.setAdapter(mAdapter);

        mBinding.viewEmptyData.setVisibility(View.GONE);
        mBinding.progressBar.setVisibility(View.VISIBLE);

        mBinding.toolbar.setNavigationOnClickListener(v -> onBackPressed());

        mBinding.btnClearData.setOnClickListener(this);
        mBinding.btnGetData.setOnClickListener(this);
    }

    private void checkAndShowEmptyView() {
        if (mAdapter != null) {
            if (mAdapter.isEmpty()) {
                mBinding.viewEmptyData.setVisibility(View.VISIBLE);
                mBinding.viewEmptyData.showEmptyAd();
            } else {
                mBinding.viewEmptyData.setVisibility(View.GONE);
            }
        }
    }

    @Override
    public void showData(List<String> data) {
        mBinding. progressBar.setVisibility(View.GONE);
        mAdapter.setData(data);
        checkAndShowEmptyView();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_get_data:
                mBinding. progressBar.setVisibility(View.VISIBLE);
                mPresenter.getData();
                break;
            case R.id.btn_clear_data:
                mPresenter.clearData();
                break;
        }
    }
}

File: app/src/main/java/com/tohsoft/app/ui/history/HistoryActivity.kt
package com.tohsoft.app.ui.history

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.SearchView
import androidx.paging.DataSource
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import androidx.recyclerview.widget.LinearLayoutManager
import com.tohsoft.app.R
import com.tohsoft.app.data.models.History
import com.tohsoft.app.databinding.ActivityHistoryBinding
import com.tohsoft.base.mvp.ui.BaseActivity
import com.utility.DebugLog
import java.util.*

/**
 * Created by PhongNX on 2/10/2020.
 *
 * Template cho việc hiển thị list data (hiển thị emptyAd...)
 * Note: Nếu không dùng Activity này thì xóa đi
 */
class HistoryActivity : BaseActivity(), View.OnClickListener {
    private lateinit var mBinding: ActivityHistoryBinding
    private lateinit var mViewModel: HistoryViewModel
    private var mAdapter: HistoryPagingAdapter? = null
    private var mAdapterSearched: HistoryPagingAdapter? = null

    override val bottomAdsContainer: ViewGroup
        get() = mBinding.frBottomBanner

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = ActivityHistoryBinding.inflate(layoutInflater)
        setContentView(mBinding.root)

        init()
        registerLiveDataListener()
    }

    private fun init() {
        mViewModel = getViewModel(HistoryViewModel::class.java)

        mAdapter = HistoryPagingAdapter()
        mBinding.rvData.layoutManager = LinearLayoutManager(mContext)
        mBinding.rvData.adapter = mAdapter

        mBinding.viewEmptyData.visibility = View.GONE
        mBinding.progressBar.visibility = View.VISIBLE
        mBinding.toolbar.setNavigationOnClickListener { onBackPressed() }
        mBinding.btnClearData.setOnClickListener(this)
        mBinding.btnGetData.setOnClickListener(this)

        mBinding.searchView.setOnQueryTextListener(object : androidx.appcompat.widget.SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                mViewModel.searchHistory(query)
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                mViewModel.searchHistory(newText)
                return true
            }
        })

        mBinding.searchView.setOnCloseListener {
            mBinding.searchView.onActionViewCollapsed()
            mBinding.rvData.adapter = mAdapter
            false
        }

        mBinding.searchView.setOnQueryTextFocusChangeListener { v, hasFocus ->
            if (mBinding.rvData.adapter == mAdapter) {
                mAdapterSearched = HistoryPagingAdapter()
                mBinding.rvData.adapter = mAdapterSearched
            }
        }
    }

    private fun registerLiveDataListener() {
        mViewModel.progressState.observe(this, { visible ->
            mBinding.progressBar.visibility = if (visible) View.VISIBLE else View.GONE
        })
        mViewModel.historiesPaged.observe(this, { data ->
            showData(data)
        })
        mViewModel.historiesSearchPaged.observe(this, { data ->
            showSearchedData(data)
        })
    }

    private fun checkAndShowEmptyView(data: PagedList<History>?) {
        if (data == null || data.isEmpty()) {
            mBinding.viewEmptyData.visibility = View.VISIBLE
            mBinding.viewEmptyData.showEmptyAd()
        } else {
            mBinding.viewEmptyData.visibility = View.GONE
        }
    }

    private fun showSearchedData(data: PagedList<History>?) {
        DebugLog.loge("showSearchedData ${data?.size}")
        mAdapterSearched?.submitList(data)
        checkAndShowEmptyView(data)
    }

    private fun showData(data: PagedList<History>?) {
        DebugLog.loge("showData ${data?.size}")
        mAdapter?.submitList(data)
        checkAndShowEmptyView(data)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_get_data -> mViewModel.initData()
            R.id.btn_clear_data -> mViewModel.clearData()
        }
    }
}